metrics/src/key.rs

```diff
@@ -175,7 +175,34 @@ fn generate_key_hash(name: &KeyName, labels: &Cow<'static, [Label]>) -> u64 {
 
 fn key_hasher_impl<H: Hasher>(state: &mut H, name: &KeyName, labels: &Cow<'static, [Label]>) {
     name.0.hash(state);
-    labels.hash(state);
+    labels.len().hash(state);
+    match labels.len() {
+        0 => {}
+        1 => labels[0].hash(state),
+        2 => {
+            if labels[0] < labels[1] {
+                labels[0].hash(state);
+                labels[1].hash(state);
+            } else {
+                labels[1].hash(state);
+                labels[0].hash(state);
+            }
+        }
+        n if n < 8 => {
+            let mut labels_sort_map: [u8; 8] = [0, 1, 2, 3, 4, 5, 6, 7];
+            labels_sort_map[..n].sort_by_key(|i| labels[*i as usize].key());
+            for i in 0..n {
+                labels[labels_sort_map[i] as usize].hash(state)
+            }
+        }
+        n => {
+            let mut labels_sort_map: Vec<usize> = (0..n).collect();
+            labels_sort_map.sort_by_key(|i| labels[*i].key());
+            for i in 0..n {
+                labels[labels_sort_map[i]].hash(state)
+            }
+        }
+    }
```

Comment:
I wonder if there's a way we could wrap this loop unrolling into a helper function or something, since we effectively repeat it three times: here, the `PartialEq` impl, and the `Ord` impl.

I guess the biggest thing is that this impl only deals with one set of labels, while the other two trait impls deal with a pair of labels.

Reply:
> I wonder if there's a way we could wrap this loop unrolling into a helper function or something, since we effectively repeat it three times: here, the PartialEq impl, and the Ord impl.

I can't come up with a good way of doing it.